### DISCUSSION TITLE
Support/ Reference for rseq/sseq (and other related Nintendo sequence) complex instructions

### DISCUSSION THREAD

**Discussion by bhospelhorn** [vgmtrans/vgmtrans#220] [Music formats]
It looks like it has been brought up before that the weird "machine language" like commands for DS sseq don't appear to be implemented yet  [#41] [#39].  Though in most cases for midi conversion this isn't an issue, as someone pointed out, it can break major songs like the Spirit Tracks overworld theme (which I too have tried to extract to no avail).

Instead of nagging yall again for the feature in VGMTrans (though that would be pretty neat), I wanted to instead ask about format references or documentation. Generally, when I want to write my own parser for these formats in the event that a format isn't fully supported by VGMTrans, I still use your source code as a reference (among several) on how to read the file format. **However, the switch blocks for the reading and implementation of those complex sseq commands were left blank in NDSSeq.cpp**, presumably for later implementation. My search terms are probably wrong, but every other source I find on rseq/sseq refers back to you or loveemu's sseq2mid! The rseq2midi source file is also fantastic, but it too skips over `0xan` and `0xf0` commands.
I'm sure there's a document out there that lays out the rest of the commands, but I can't find it.

My broad idea would be to add a doc section to the repo directly, so you could in theory share your sources and knowledge regarding formats or features that are yet to be implemented - of course that's probably more trouble than it's worth. Instead, I have found threads on this Issues section to be invaluable for this purpose (such as the case with the Pokemon Mystery Dungeon EoT/EoD SMD format [#38]). I figured that if nothing else, it might be handy to start a public discussion on this topic. I do apologize if I end up causing a clutter!

-------------------------------------------

I am working on a set of parser/converter (or extraction) classes in Java for the Wii BRSAR type and its most common component types (BRWAR, BRWSD, BRWAV, BRSTM, BRBNK, and BRSEQ). The file formats themselves are pretty straightfoward, but the two big obstacles are decompressing the ADPCM data (plenty of references for that, should be good there) and interpreting the rseq commands.

RSEQ seems to be basically the same as SSEQ, but with some of the command bytes changed (I wouldn't be surprised if this is the case for the default format for other Nintendo consoles as well). As such, docs and source code on SSEQ have been extremely helpful. Still, none of the sources I've used have fully implemented or outlined the `0xan` (called "Prefix" commands in BrawlBox) or `0xf0` (`0xbn` in SSEQ(?)) commands.

My question is this: **What is it exactly that each of these commands do, and how are their parameters read?**

Based on the presence of other commands like "call", "jump", and "return," I'm inclined to think they work more like traditional CPU instructions - ie. take variable in register x, modify it with the variable in register y and store the result in register z - allowing for the same sequence data to be played in different ways depending upon the state of the player. Even generously assuming that I have that basic logic correct, I still have no idea how to interpret the parameters followed by these commands.

- What references did you use to determine what these sseq commands were? 
- Do you know of any good references for sseq or rseq (or related sequence formats)? 
- If you determined the meaning and way to read these commands experimentally, what do you know?

(Of course these questions are open to anyone who is interested or might have an answer, not just VGMTrans contributors)

**Comment by sykhro** [comment]
> However, the switch blocks for the reading and implementation of those complex sseq commands were left blank in NDSSeq.cpp

I believe this was done because there isn't a nice way to handle this with the current VGMTrans "API", which is one of the things I'm trying to address in my Qt port/refactoring attempt (without much success so far, because the codebase is big and not exactly modern)

> I'm sure there's a document out there that lays out the rest of the commands, but I can't find it.

afaik there's no complete docs about the standard sdat format - kiwi's overview is probably the one containing most details. [kode's player](https://github.com/kode54/SSEQPlayer) is also a good resource for commands (maybe something like this exists in the official sdk, but we do blackbox testing just as any other reverse engineering effort) 

> My broad idea would be to add a doc section to the repo directly, so you could in theory share your sources and knowledge regarding formats or features that are yet to be implemented - of course that's probably more trouble than it's worth. 

We really need something like this, because info on most vgm formats is slowly disappearing from the internet. I'd be up to make the wiki public-writable again and have dedicated sections there.
Or maybe just comment our source code properly...

**Comment by bhospelhorn** [comment]
Sorry for the late reply. Thanks a million - kode's SSEQ player code was very helpful for the problem at hand. Still can't figure out what the a3, a4, and a5 commands do exactly, but I've tripped across them in some of the Skyward Sword SE files so maybe I can figure it out from there.

Yeah, I can totally see that being an issue. I tested a draft of my code on the Spirit Tracks overworld theme (as I mentioned) and found that these types of sequences required manual inspection to determine how to preset the variables to rip the data correctly. It was also using jump commands as loops instead of the loop start/ loop end instructions which was obnoxious and made the loop points harder to detect. All that considered, I can see why it would be a pain to try to automate midi conversion for these files in the first place, let alone trying to integrate such into existing code that was written with more midi-like sequences in mind.

I'm making a point of writing documentation for the BRSAR formats as I go through them. The best docs I could find were on the Mario Kart Wii modding wiki, but they can be somewhat scant for some formats.

**Comment by SamusAranX** [comment]
Hiya!
I just found my way here after (unsuccessfully) trying to rip the Spirit Tracks overworld theme and taking to Google after I couldn't find the intro part.
I'm curious, has any progress been made on this? 
Excuse my ignorance (most of this goes way over my head), but: In lieu of a fancy way to deal with variables and jumps, would it be feasible to just dump everything into a MIDI file and leave it to the user to separate track parts in a DAW?

**Comment by bhospelhorn** [comment]
I don't think VGMTrans has added a way to handle these yet. I did write a seq ripper for myself that does work on the SpiritTracks overworld theme (https://github.com/waffleoRai/NinSeq-Viewer) - it was mostly for myself; I use github primarily to transfer stuff between computers so the documentation's a bit crap. If you are interested in using it, I can make it more user friendly too :)
I've also taken to writing up some informal docs on various formats I familiarize myself with - especially since Nintendo's are weird and they recycle them like crazy.

Unfortunately, the issue is that VGMTrans won't dump most of the song data to a MIDI file at all (though if you can GET it into a MIDI, you can import it into just about any DAW). What happens is that when VGMTrans hits an "if" instruction, it doesn't check the condition flag that should have been set by a previous instruction - it just always assumes it's true(I think, will have to double check, might be false... anyway, it doesn't evaluate the condition). The Spirit Tracks overworld theme uses these "if" instructions liberally both as repeat signs and to determine where to enter the song based on what is happening in the game. A very early "if" check in the sequence looks at one of the state variables to determine where in the song to start. I think the issue was that when this "if" is ignored, (and other jumps and function calls are not taken correctly) VGMTrans' midi converter essentially ends up jumping to the end of the track. Thus, the data are skipped, nothing gets converted, and the midi is almost empty.

Now, in theory - if you could get a linear dump of the data (ie. ignore jumps and function calls and branches and just dump the timed notes as is) you COULD chop up and move around the musical phrases in a DAW that supports midi sequences to reconstruct the song. Unfortunately, VGMTrans isn't wired to do a linear dump like that at the moment. Those branches really mess up how it reads the file.

What I did for my tool is I essentially wrote my converter to behave like an emulator of sorts. You always have to watch out for those infinite jump loops, but other than that the MIDI converter just generates MIDI instructions and outputs them to a stream based on whatever the SSEQ reader is telling it to do (this is because SSEQ logic is more akin to machine language while MIDI is more streamlike). Because SSEQ to some extent operates under a different execution model than MIDI, only the basic music instructions can be translated one for one - that's why the usual conversion approach that works for most sequence formats is a bit of  a crapshoot when Nintendo decides to go balls out and utilize the full extent of the format's capabilities for a song.

I sort of rambled on there - I hope that makes some sense! I think the takeaway is this - if you want to write your own converter, you'll unfortunately have to learn the quirks of MIDI vs. Nintendo's *SEQ formats. If it's just the MIDI conversion you are interested in, I'm sure I'm not the only one who wrote a tool explicitly to handle a SSEQ like this (there may be forks of VGMTrans that have patched this too...). Afaik the current main VGMTrans version has this data skip issue during conversion and I don't know if it's a priority for the team to fix it since I think it's really a problem specific to this format family and even then, not an issue for a lot of songs that do use that format (dumped the seqs from all the pokemon ds games with nary a complaint).

**Reply by mguivh** [reply]
Hi! I stumbled over this page while trying to rip the Spirit Tracks overworld theme, and your seq ripper seems perfect except I don’t understand how to use it. I don’t know how to find out which variable values should be entered (when I don’t change anything the MIDI only contains one short note). I might also be doing everything else wrong. Any help/instructions would be greatly appreciated!

**Reply by waffleoRai** [reply]
Hi! Sorry for the delayed reply! It's been a hot second since I last used that particular GUI so I'll have to take a look at it. I thought that I had a little form for entering initial variable values, but I'm not 100% sure on that one. I'd also have to take another quick look at the sseq command list because I did have a setting I used for that track (I think any value > 16 for one of the variables would do the trick - force it to go through all the phrases instead of skipping them all. But I don't remember which variable slot offhand.)
I'll give it a look when I get home today and let you know : )

**Reply by waffleoRai** [reply]
Hi again! I looked it over and basically yeah - the GUI SHOULD allow you to set behavior and initial variable values for MIDI export (I just put it in a stupid place - it's in MIDI -> MIDI Conversion Settings...)
The overworld theme uses variables 0 and 1 to track what section the song is in (and where it last left off) and where it should jump to should the sequence be reloaded and restarted. It has 16 jump points determined by whether variable 0 holds the value 0-15 when the sequence reaches the entrypoint fork (on track 0 this starts at 0x47). If any other value is in slot 0, then the sequence just runs through the branches without jumping and starts at the proper song intro. 

In other words, to get the player/MIDI exporter to read the full song, you need to start the read with the variable in slot 0 set to anything OTHER than 0-15. I usually just set it to 16 and that does the trick.

In my viewer GUI you should be able to do this by going to the MIDI top menu, clicking the MIDI Conversion Settings... option, going down to the last option which is "Initial Variable Values", typing "16" into the text box (making sure the spinner box is set to 0), hitting "save", then "confirm", then going to MIDI->Export to MIDI...

Let me know if you give it a whirl and if you have any more issues. There may well be other bugs that I missed.

**Reply by mguivh** [reply]
Thanks for the help! It all makes a lot more sense to me now. 
However, there still seems to be some other problem which is causing it to put out a midi file with only a single short note (I might be wrong, but I don't think it even belongs anywhere in the track). I don't think it is related to the jump points since I (only now) tried with some other tracks and got the same problem (but with different notes). I’ve wondered if the problem is with the sseq files I am using but it looks right to me untill I export…

**Reply by waffleoRai** [reply]
Hmm. Well first thing's first - I just rebuilt the JAR and pushed that earlier today. I didn't change any of the GUI source code, but I've made a LOT of changes to the util library backing the GUI (which includes all the format handlers) since 2019. So try that one instead just to make sure we're running the same version. Also I just wanna double check to make sure we are looking at the EXACT same sseq file (we should be, but sometimes weird things happen in extraction) so I'm including a screencap of some checksums on a copy of the file I extracted.

I'll try to see if I can recreate the issue... I do vaguely remember this being a problem for me at one point and I think there must be some jump or if behavior that's not set correctly, but I don't remember what does that.

![image](https://user-images.githubusercontent.com/24638992/135200215-76476b0d-f2a8-48dd-951f-e9aa47dd44bb.png)

**Reply by mguivh** [reply]
I tried the new JAR and this time it worked immediately, whithout doing anything else differently. Thanks again for all the help!

**Reply by waffleoRai** [reply]
Awesome! No problem! Love it when it ends up being that simple.

**Comment by sykhro** [comment]
Not a whole lot is planned feature-wise right now; I'm (still) taking care of refactoring, the rest of development is spurious bugfixes. For the record, PRs are welcome in case anyone wants to try and implement a workaround
